### PR TITLE
Weierstrass Preparation Theorem 

### PR DESCRIPTION
**Pull request by Gebi95**
This is my first pull request. It is a porting from the weierstr.lib from Singular. 

(Caution: Weierstrass Preparation Theorem and Weierstrass Division Theorem are not related to Weierstrass Form of Elliptic curves)

### CODE REVIEW

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+		else
+			i=1
+			while i<=j 
+				if i!=n
+					m[i]=gens(R)[i]+gens(R)[n]^rand(2:voice*d)
+				end	
+				i=i+1
+			end
+			g=evaluate(f,m)
+			return var_general(g,voice+1)
```
Comment (lkastner) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]: These two blocks look identical except for the line `m[i]=...` and the return statement. Any chance of refactoring this two make a small function for the inner loop?

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+transform `f` into a polynomial, which is x_n general of finite order for x_n the `n`-th ring variable.
+"""
+function var_general(f::MPolyElem,n::Int)
+	return var_general(f,n,0)
+end
+
+"""
+`var_general(f::MPolyElem,n::Int,voice::Int)`
+
+transform `f` into a polynomial, which is x_n general of finite order for x_n the `n`-th ring variable. `voice` is the counter of recursive calls (like nesting_depth). First call with `voice`=0
```
Comment (lkastner) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]: If `voice` "is the counter of recursive calls (like nesting_depth)", why not call it `nesting_depth` immediately? `voice` definitely is a non-standard name and makes the code harder to read.

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+#
+#Conversion from Singular (weierstr.lib)
+#
+#############################################################
+
+"""
+`order_of_generality(f::MPolyElem)`
+
+compute for a given polynomial `f` an integer `d` such that `f` is `x_n`-general of order `d` with `x_n` the n-th variable of the polynomial ring of `f` 
+"""
```
Comment (lkastner) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]: The docs need to be brushed up a bit, please have a look at
https://docs.oscar-system.org/dev/DeveloperDocumentation/documentation/
Also please wrap the lines to max 90-100 chars, I think that is the number we have in the style guide.

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+end
+
+"""
+`last_var_general(f::MPolyElem,n::Int)`
+
+transform `f` into a polynomial, which is x_n general of finite order for x_n the `n`-th ring variable.
+"""
+function var_general(f::MPolyElem,n::Int)
+	return var_general(f,n,0)
+end
```
Comment (lkastner) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]:
```suggestion
var_general(f::MPolyElem,n::Int) = var_general(f,n,0)
```

**Review by lkastner — COMMENTED** [review]
A few general points, probably you already have them on your agenda, just to make sure:
- There are no tests yet.
- Nothing is exported yet.
- Maybe add some spaces here and there? E.g. `f = evaluate(...)` instead of `f=evaluate(...)`. Also after `,` for separating arguments.
- Have a look at our [styleguide](https://docs.oscar-system.org/dev/DeveloperDocumentation/styleguide/)
- Use longer variable names where it makes sense. E.g. `degree` instead of `de`, or `degree_f` if you are also using the function `degree`. This makes it easier to read and to grep for stuff.
- I think you can discard stuff from a function outputting a tuple by using `_` if you are not interested in that variable.

Thanks!

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+Given a multivariate polynomial `f` and an integer degree `de`, performing the weierstrass preparation theorem of `f` up to degree `de`. Returning an unit `u` and a weierstrass polynomial `fhat`, s.t. `u``fhat`=`f`, while `fhat` is weierstrass. Additional the number of iterations used will be returned. 
+"""
+function weierstrass_preparation_theorem(f::MPolyElem,de::Int,n::Int)
+	
+	if f==0
+		throw(DomainError(f, "polynomial is zero"))
+	end
+	R=parent(f)
+	success,d=order_of_generality(f,n)
+	if success!=true
```
Comment (lkastner) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]:
```suggestion
	if !success
```
`success` is a bool, right?

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+`weierstrass_division_theorem(g::MPolyElem,f::MPolyElem,de::Int)`
+
+Given a multivariate polynomial `f`, which is general of oder `d` in the last ring variable, a polynomial `g` and an integer degree `de`, performing the weierstrass division theorem of `f` up to degree `de`, by dividing `f` by `g`. Returning a polynomial `u` and a remainder `r` up to degree `de`, s.t. f=ug+r. Additional the number of iterations will be returned. 
+"""
+function weierstrass_division_theorem(g::MPolyElem,f::MPolyElem,de::Int,n::Int)
+	R=parent(f)
+	r=R(0)
+	vect=fill(0,nvars(R))
+	vect[n]=1 
+	success,d=order_of_generality(f,n)
```
Comment (lkastner) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]: It seems like `success` is never used, if I am not mistaken. Then could you just do
```suggestion
	_,d = order_of_generality(f,n) 
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl (hunk 1 above)
Comment (Gebi95) [reply experimental/Rings/WeierstrassPreparationTheorem.jl]: Thanks for your comments. 

I tried to solve this by creating the function rand_coeff, but i am not sure, whether it is a good idea to have an operator as an argument for this function?

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 7]
```diff
@@ -0,0 +1,240 @@
+##############################################################
+#
+#Conversion from Singular (weierstr.lib)
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]:
```suggestion
# Conversion from Singular (weierstr.lib)
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+#
+#############################################################
+
+export weierstrass_preparation_theorem
+
+export weierstrass_division_theorem
+
+
+"""
+`order_of_generality(f::MPolyElem, n::Int)`
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]: Please use exactly four spaces to indent the signature(s) in docstrings, and use no extra backticks around it
```suggestion
    order_of_generality(f::MPolyElem, n::Int)
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+compute for a given polynomial `f` an integer `d` such that `f` is `x_n`-general of order `d`
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]: General rules of grammar apply for docstrings, too (so e.g. sentences are capitalized and end with a period, question mark or exclamation mark),
```suggestion
Compute for a given polynomial `f` an integer `d` such that `f` is `x_n`-general of order `d` 
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+export weierstrass_preparation_theorem
+
+export weierstrass_division_theorem
+
+
+"""
+`order_of_generality(f::MPolyElem, n::Int)`
+
+compute for a given polynomial `f` an integer `d` such that `f` is `x_n`-general of order `d` 
+with `x_n` the n-th variable of the polynomial ring of `f`
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]:
```suggestion
with `x_n` the `n`-th variable of the polynomial ring of `f`.
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl (hunk 9 above)
Comment (fingolfin) [reply experimental/Rings/WeierstrassPreparationTheorem.jl]: What does `` `x_n`-general `` mean?

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+"""
+`order_of_generality(f::MPolyElem, n::Int)`
+
+compute for a given polynomial `f` an integer `d` such that `f` is `x_n`-general of order `d` 
+with `x_n` the n-th variable of the polynomial ring of `f` 
+"""
+function order_of_generality(f::MPolyElem, n::Int)
+	 
+	vect = fill(parent(f)(0), nvars(parent(f))) 
+	vect[n] = gens(parent(f))[n]
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]:
```suggestion
	vect[n] = gen(parent(f), n)
```
`gen(R,i)` is more efficient than `gens(R)[i]`; so this should be also done in other places (e.g. a few lines below in this function)

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+"""
+`jet(f::MPolyElem, degree_f::Int)`
+
+Given a polynomial `f` and an integer degree `degree_f` , this function returns the polynomial,
+which contains only the monomials of `f`, which total degree is lower than or equal to `degree_f`
+"""
+function jet(f::MPolyElem, degree_f::Int) 
+	while f != 0
+		if total_degree(leading_monomial(f, ordering=deglex(gens(parent(f))))) > degree_f
+			f = f-coeff(f, leading_monomial(f, ordering=deglex(gens(parent(f)))))*leading_monomial(f, ordering=deglex(gens(parent(f))))
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]: It seems you are computing `leading_monomial(f, ordering=deglex(gens(parent(f))` three times here? Better to store it in a variable... and also the ordering you keep re-creating...

```suggestion
    ordering = deglex(gens(parent(f)))
	while !iszero(f)
	    lm = leading_monomial(f, ordering)
		if total_degree(lm) > degree_f
			f = f - coeff(f,lm)*lm
```

That said, I am pretty sure this code is very inefficient in its current form. What you really want is to iterate over all *terms* and discard those whose total degree is too big. That does not require a specific monomial ordering; in particular the "intrinsic" ordering of the polynomial implementation could be used...

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+transform `f` into a polynomial, which is x_n general of finite order for x_n the `n`-th ring variable.
+`nesting_depth` is the counter of recursive calls. First call with `nesting_depth` = 0
+"""
+function var_general(f::MPolyElem, n::Int, nesting_depth::Int) 
+	R = parent(f)
+	j = nvars(R)
+	_, d = order_of_generality(f, n) 
+	if d > 0 
+		return f	
+	else
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl]: The more indented a function, the harder it gets to understand it. Here, you can save one level: just do
```suggestion
	end
```
and then de-indent the following code, and finally drop the closing `end`

**Review by fingolfin — CHANGES_REQUESTED** [review]
Thanks for your contribution!

I've left a couple comments. Please apply them in all places where it makes sense

File: experimental/Rings/WeierstrassPreparationTheorem.jl (hunk 12 above)
Comment (Gebi95) [reply experimental/Rings/WeierstrassPreparationTheorem.jl]: My idea was, to use this ordering to not iterate over all terms of f. But maybe this is inefficient. So i changed it now to 
  i = length(f)
  while i >= 1
    if total_degree(term(f,i)) > degree_f
      f = f - term(f,i)
    end
    i = i - 1
  end

I hope, this matches your idea.

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+	vect[n] = gen(parent(f),n) 
+	f = evaluate(f, vect) 
+	f != 0 || return false, -1
+	return true, degree(leading_monomial(f, ordering=neglex(gens(parent(f)))), n)
+end
+
+"""
+    jet(f::MPolyElem, degree_f::Int)
+
+Given a polynomial `f` and an integer degree `degree_f` , this function returns the polynomial,
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:30]:
```suggestion
Given a polynomial `f` and an integer degree `degree_f`, this function returns the polynomial
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+which contains only the monomials of `f`, which total degree is lower than or equal to `degree_f`.
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:31]:
```suggestion
which contains only the monomials of `f` whose total degree is lower than or equal to `degree_f`.
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+"""
+function jet(f::MPolyElem, degree_f::Int) 
+  i = length(f)
+  while i >= 1
+    if total_degree(term(f,i)) > degree_f
+      f = f - term(f,i)
+    end
+    i = i - 1
+  end 
+	return f
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:41]:
```suggestion
  return f
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+"""
+    jet(f::MPolyElem, degree_f::Int, vecw::Vector{Int})
+
+Given a polynomial `f`, an integer degree `degree_f` and an integer vector `vecw`, which has to have 
+the same length as length(parent(f)), this function returns a polynomial, which only contains 
+monomial, whose exponents taken as a vector multiplied (scalar product) with the given 
+weight-vector `intw` are lower than or equal to `degree_f`.
+"""
+function jet(f::MPolyElem, degree_f::Int, vecw::Vector{Int}) 
+
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:54]:
```suggestion
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+end
+
+
+"""
+    jet(f::MPolyElem, degree_f::Int, vecw::Vector{Int})
+
+Given a polynomial `f`, an integer degree `degree_f` and an integer vector `vecw`, which has to have 
+the same length as length(parent(f)), this function returns a polynomial, which only contains 
+monomial, whose exponents taken as a vector multiplied (scalar product) with the given 
+weight-vector `intw` are lower than or equal to `degree_f`.
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:51]:
```suggestion
weight-vector `vecw` are lower than or equal to `degree_f`.
```
(Although the `vec` is somewhat redundant, so perhaps just call it `w` or `weights`?)

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+"""
+function jet(f::MPolyElem, degree_f::Int, vecw::Vector{Int}) 
+
+	vect = collect(exponent_vectors(f))
+	i = length(f)
+	while i >= 1
+		if transpose(vecw)*vect[i] > degree_f
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:58]:
```suggestion
		if dot(vecw,vect[i]) > degree_f
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+			f = f-term(f, i)
+		end
+		i = i-1	
+	end
+	return f
+end
+
+"""
+    invert_unit(u::MPolyElem, n::Int)
+If `u` is an unit, returning boolean value for success and the inverse of `u`.
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:68]:
```suggestion

If `u` is an unit, returning boolean value for success and the inverse of `u`. 
```

And what is `n` doing?

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+	return f
+end
+
+"""
+    invert_unit(u::MPolyElem, n::Int)
+If `u` is an unit, returning boolean value for success and the inverse of `u`. 
+Else return false and the zero polynomial. 
+"""
+function invert_unit(u::MPolyElem, n::Int)
+	if u != 0 && total_degree(leading_monomial(u, ordering=negdegrevlex(gens(parent(u))))) == 0
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:72]: Can't you also use this equivalently?
```suggestion
	if u != 0 && total_degree(leading_monomial(u, ordering=:negdegrevlex)) == 0
```

Comment (fingolfin) [reply experimental/Rings/WeierstrassPreparationTheorem.jl:72]: Is this code perhaps assuming that the coefficient ring is a field?

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+    var_general(f::MPolyElem, n::Int, nesting_depth::Int)
+
+Transform `f` into a polynomial, which is x_n general of finite order for x_n the `n`-th ring variable.
+`nesting_depth` is the counter of recursive calls. First call with `nesting_depth` = 0.
+"""
+function var_general(f::MPolyElem, n::Int, nesting_depth::Int) 
+	R = parent(f)
+	j = nvars(R)
+	_, d = order_of_generality(f, n) 
+	if d > 0
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:109]: Please replace all tabs by spaces

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+Transform `f` into a polynomial, which is x_n general of finite order for x_n the `n`-th ring variable.
+`nesting_depth` is the counter of recursive calls. First call with `nesting_depth` = 0.
+"""
+function var_general(f::MPolyElem, n::Int, nesting_depth::Int) 
+	R = parent(f)
+	j = nvars(R)
+	_, d = order_of_generality(f, n) 
+	if d > 0 
+		return f	
+  end
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:111]: Indention is wrong here (probably a tabs versus spaces issue?)

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+			if length(g) > length(evaluate(g, m_hilf))
+				m[i] = gen(R,i)+rand(1-(nesting_depth)*10 : 1+(nesting_depth)*10)*gen(R,n)
+				g = evaluate(f, m)
+				break
+			end
+		end
+		i = i+1
+		end
+		
+		if nesting_depth <= 3
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:128]: Indentation is definitely way off

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+function initial_form(f::MPolyElem)
+	degree_f = total_degree(leading_monomial(f, ordering=negdegrevlex(gens(parent(f))))) 
+	g = jet(f, degree_f) 
+	return g
+end
+
+"""
+    rand_coeff(m::Vector{fmpq_mpoly}, n::Int, mini::Int, maxi::Int, operator)
+
+Given a vector of monomials `m`, an Integer `n`, integer bounds mini und maxi and an
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:155]:
```suggestion
Given a vector of monomials `m`, integer `n`, `mini` and `maxi`, and an 
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+which is general for the `n`-th ring variable.
+"""
+function rand_coeff(m::Vector{fmpq_mpoly}, n::Int, mini::Int, maxi::Int, operator)
+	R=parent(m[1])
+	j=nvars(R)
+	i = 1
+	while i <= j 
+		if i != n
+			m[i] = gen(R,i)+operator(gen(R,n),rand(mini : maxi))
+		end
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:166]: maybe pass the range `mini:maxi`?

File: experimental/Rings/WeierstrassPreparationTheorem.jl (hunk 1 above)
Comment (lkastner) [reply experimental/Rings/WeierstrassPreparationTheorem.jl]: I think that is ok, we have operators as arguments elsewhere too. Thanks.

**Review by fingolfin — CHANGES_REQUESTED** [review]

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 27]
```diff
@@ -0,0 +1,239 @@
...
+Compute for a given polynomial `f` an integer `d` such that `f` is `x_n`-general of order `d` 
+with `x_n` the `n`-th variable of the polynomial ring of `f`.
+"""
+function order_of_generality(f::MPolyElem, n::Int)
+	 
+	vect = fill(parent(f)(0), nvars(parent(f))) 
+	vect[n] = gen(parent(f),n) 
+	f = evaluate(f, vect) 
+	f != 0 || return false, -1
+	return true, degree(leading_monomial(f, ordering=neglex(gens(parent(f)))), n)
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:24]:
```suggestion
	R = parent(f)
	vect = fill(R(0), nvars(R)) 
	vect[n] = gen(R,n) 
	f = evaluate(f, vect) 
	f != 0 || return false, -1
	return true, degree(leading_monomial(f, ordering=neglex(gens(R))), n)
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl [hunk 28]
```diff
@@ -0,0 +1,239 @@
...
+export weierstrass_preparation_theorem
+
+export weierstrass_division_theorem
+
+
+"""
+    order_of_generality(f::MPolyElem, n::Int)
+
+Compute for a given polynomial `f` an integer `d` such that `f` is `x_n`-general of order `d` 
+with `x_n` the `n`-th variable of the polynomial ring of `f`.
```
Comment (fingolfin) [inline experimental/Rings/WeierstrassPreparationTheorem.jl:16]:
```suggestion
Return an integer `d` such that `f` is ``x_n``-general of order `d` 
with ``x_n`` the `n`-th variable of the polynomial ring of `f`.
```

File: experimental/Rings/WeierstrassPreparationTheorem.jl (hunk 9 above)
Comment (afkafkafk13) [reply experimental/Rings/WeierstrassPreparationTheorem.jl]: Textbook definition:
A power series f \in CC{x_1,..,x_n} is called x_n-general of order d if f(0,..,0,x_n) has a zero of order d at x_n=0.

File: experimental/Rings/WeierstrassPreparationTheorem.jl (hunk 21 above)
Comment (afkafkafk13) [reply experimental/Rings/WeierstrassPreparationTheorem.jl:72]: In the context of the Weierstrass Preparation theorem, I would assume that the coefficient ring is technically QQ and theoretically CC.

**Comment by fingolfin** [comment]
What's the plan here? There are a bunch of comments that are not yet addressed.

**Comment by afkafkafk13** [comment]
I did a few tests with this implementation and it is far from optimal; I doubt that it is worth adressing the comments, before its major flaws have been addressed.  We will discuss with the author whether this will be closed or refactored.

**Comment by afkafkafk13** [comment]
Refactoring will not take place in this PR.

**Pull request closed without merging**